qiskit/transpiler/passes/optimization/optimize_annotated.py

```diff
@@ -0,0 +1,188 @@
+from qiskit.transpiler.target import Target
+from qiskit.dagcircuit import DAGCircuit
+from qiskit.transpiler.exceptions import TranspilerError
+
+
+class OptimizeAnnotated(TransformationPass):
+    """Optimization pass on circuits with annotated operations."""
+
+    def __init__(
+        self,
+        target: Optional[Target] = None,
+        equivalence_library: Optional[EquivalenceLibrary] = None,
+        basis_gates: Optional[List[str]] = None,
+    ):
+        """
+        OptimizeAnnotated initializer.
+
+        Args:
+            target: Optional, the backend target to use for this pass.
+            equivalence_library: The equivalence library used
+                (instructions in this library will not be optimized by this pass).
+            basis_gates: Optional, target basis names to unroll to, e.g. `['u3', 'cx']`
+                (instructions in this list will not be optimized by this pass).
+                Ignored if ``target`` is also specified.
+        """
+        super().__init__()
+
+        self._target = target
+        self._equiv_lib = equivalence_library
+        self._basis_gates = basis_gates
+
+        self._top_level_only = self._basis_gates is None and self._target is None
+
+        if not self._top_level_only and self._target is None:
+            basic_insts = {"measure", "reset", "barrier", "snapshot", "delay"}
+            self._device_insts = basic_insts | set(self._basis_gates)
+
+    def run(self, dag: DAGCircuit):
+        """Run the OptimizeAnnotated pass on `dag`.
+
+        Args:
+            dag: input dag.
+
+        Returns:
+            Output dag with higher-level operations optimized.
+
+        Raises:
+            TranspilerError: when something goes wrong.
+
+        """
+        dag, _ = self._run_inner(dag)
+        return dag
+
+    def _run_inner(self, dag) -> Tuple[DAGCircuit, bool]:
+        """
+        Optimizes annotated operations.
+        Returns True if did something.
+        """
+        # Handle control-flow
+        for node in dag.op_nodes():
+            if isinstance(node.op, ControlFlowOp):
+                node.op = control_flow.map_blocks(self.run, node.op)
+
+        # First, optimize every node in the DAG.
+        dag, opt1 = self._canonicalize(dag)
+
+        opt2 = False
+        if not self._top_level_only:
+            # Second, recursively descend into definitions.
+            # Note that it is important to recurse only after the optimization methods have been run,
+            # as they may remove annotated gates.
+            dag, opt2 = self._recurse(dag)
+
+        return dag, opt1 or opt2
+
+    def _canonicalize(self, dag) -> Tuple[DAGCircuit, bool]:
+        """
+        Combines recursive annotated operations and canonicalizes modifiers.
+        Returns True if did something.
+        """
+
+        did_something = False
+        for node in dag.op_nodes():
+            if isinstance(node.op, AnnotatedOperation):
```

Comment:
```suggestion
        for node in dag.op_nodes(op=AnnotatedOperation):
```

Reply:
I should remember to apply this pattern when possible. Done in [d724aaa](https://github.com/Qiskit/qiskit/pull/11476/commits/d724aaa54917a8fb71d3ffa49b5f0404b339aae5).